lib/spack/spack/audit.py

```diff
@@ -798,3 +801,75 @@ def _analyze_variants_in_directive(pkg, constraint, directive, error_cls):
             errors.append(err)
 
     return errors
+
+
+#: Sanity checks on package directives
+external_detection = AuditClass(
+    group="externals",
+    tag="PKG-EXTERNALS",
+    description="Sanity checks for external software detection",
+    kwargs=("pkgs",),
+)
+
+
+def packages_with_detection_tests():
+    """Return the list of packages with a corresponding detection_test.yaml file."""
+    import spack.config
+    import spack.util.path
+
+    # Directories where we have repositories
+    repo_dirs = [spack.util.path.canonicalize_path(x) for x in spack.config.get("repos")]
+
+    # Compute which files need to be tested
+    to_be_tested = []
+    for repo_dir in repo_dirs:
+        pattern = os.path.join(repo_dir, "packages", "**", "detection_test.yaml")
+        pkgs_with_tests = [os.path.basename(os.path.dirname(x)) for x in glob.glob(pattern)]
+        to_be_tested.extend(pkgs_with_tests)
+
+    return to_be_tested
```

Comment:
There's a _small_ performance issue. If I cycle over packages and read them things can get extremely slow. 

Below, on my laptop, that means going from `0.8s` of the current implementation to `28s`:
![Screenshot from 2023-09-04 10-42-00](https://github.com/spack/spack/assets/4199709/a8a18435-7182-415b-976e-168dedbd112f)

So this is roughly a 35x slowdown :thinking:

Reply:
See https://github.com/spack/spack/pull/18175/commits/b9c42ab5bee884f0bdaad1a911c57d9d650b69af This requires to add a file like:
```yaml
includes:
- "builtin.gcc"
```
to reuse the detection tests file for a custom GCC, but still runs in `0.8s` instead of half a minute.